### ISSUE TITLE
Implementing bidirectional flow

### ISSUE THREAD

**Issue by BaptisteFrancois** [pywr/pywr#702]
Hi, 
I am looking for a way to implement bidirectional flow. 
The idea is that thanks to a water pump, the flow through a pipeline could flow into both directions. 
I know that flow cannot be negative. I know that I could create two nodes 'link' for each direction. However, how to make sure that the LP will not send water into both links (i.e., both directions) during the same time step?

I also have a similar question for the system below:
A-----\
B------|----- D
C-----/

The idea is that water from A, B and C nodes can flow into the node D. However, given the system I try to model, when A is flowing into D, this means that flow from B and C to D must equal 0.  Similar, when B flows into D, flows from A and C must equal 0 and so forth and so forth.

Any idea?

**Comment by jetuk** [comment]
For the bi-directional flow issue try placing a small cost on using the transfer - by placing a Link node in the middle. The LP should not use the transfer in both directions at the same time in this case. I am not sure it is possible to but a hard constraint on this.

For the second setup I am also unsure if this is possible with the LP. I think it might require integer variables. I will think about it this evening.

**Comment by BaptisteFrancois** [comment]
So, any thoughts about how I could implement integer variables to solve the above issue?
Thanks!

**Comment by jetuk** [comment]
First, I am pretty sure we can't do this constraint with only a LP.  For some context, how many of these would you expect in your model? Is there a way to reformulate the network to get the same/similar behaviour without a binary constraint?

We would need to add some binary/integer variables to the allocation problem - thereby making it a MILP. This is possible with GLPK. The changes required to support it are fairly low level to Pywr however. I think we would need,
 
 1. A new flag on a node that if true (default false) would create a binary variable to control activation.
 2. In the LP formulation add a new binary column for each node with this flag set. 
 3. When creating the min/max flow constraints (rows) for these nodes formulate it as: `x - yQ < =0` instead of `x <= Q`, where `y` is the new binary variable. NB this is a simplification of `x` which can be more than 1 column.
 4. Add a new node type (or maybe adjust the existing `AggregatedNode`) that adds a constraint across these binary variables, e.g. `yA + yB + yC <= 1` would be the example above. Add support for this node/constraint in the solver.
 5. If any binaries are present use `glp_intopt` instead of `glp_simplex` to solve the problem using branch & bound.

This is fairly non-trivial but would make pywr much more flexible. Though I note that the MILP solve is likely to be much slower than standard LP.

**Comment by BaptisteFrancois** [comment]
Hi James, 
Thanks for your answer, and thanks for the suggested work plan to allow MILP in Pywr. However, I obviously a Pywr's user, meaning that I do not feel able to implement this by myself, even if this is a fairly low level change to the src code. I would be happy to test this implementation if you decide to update the current Pywr though. 

I'll try, at least for now, to reformulate the network to avoid this. My guess is that by using pre-defined rules (if water flows from A to C, then water cannot flow from B to C), but I guess we are loosing here the aim of LP. 

FYI; The network I a trying to model is actually quite complicated, at least part of it. It includes several input and output that are connected through the same network of pipelines. In reality flows are controlled by an ensemble of valves and pumps that do not allow several nodes to let water flowing into the same pipe at the same time. 
After some simplification that I think I could handle by myself, I have reduced the part of the system under consideration to the schematic below:

R --**>**-**>**---\\
H--**>**-**>**----|--- Storage node ---**>**-**>**--- Output
C---->----/

R, H, C are different nodes that can all send water to a storage node. The pipe network does not allow water from R and H to flow to the storage node at the same time. Outflow from the storage can flow to another part of the system (here called output) but uses the same pipe that is used to pump water from R and H nodes to the storage nodes, which means that R and H cannot flow to the storage when water is extracted from the storage by this mean. This is represented on the schematic above by double arrows (double arrows show water transfers that cannot be used simultaneously. 

If you have any idea how I could represent such as system by avoiding integer, I'll go for it. Maybe in some previous projects you worked on?

**Comment by jetuk** [comment]
@BaptisteFrancois sorry I didn't intend to imply that you should do this! I agree it is quite involved. However, having put some thought in to it over the weekend I thought I would write it down. We should probably make separate issue about adding MILP support. 

Re your model. Can the decision regarding use of R, H or outflow be made ahead of time. I.e. could you write a parameter that used current system state (e.g. storage volume) to decide which link was active? Or does it have to be decided as part of the LP?  If you can do the former it would make this problem a lot easier. This has been the tactic I've used for binary operating decisions in the past; they have usually been attached to a storage control curve that triggers their activation.

**Comment by BaptisteFrancois** [comment]
@jetuk no worries! 
I tried to make this conditions dependent on some states variables but this is not that simple. Through the discussions I had with the operators it looks there is no such a simple rules but more a hundreds of rules... I'll try to simplify the system.

About the way you handle this issue I am curious about the way you formulate your solution, especially the one using a control curve. Is the idea to combine, say a storagethreshold parameter with a control curve?

**Comment by BaptisteFrancois** [comment]
By the way, do you want me to open a new issue about the MILP support?

**Comment by jetuk** [comment]
Yes, or we could rename this one? I don't mind which. Label this one has an "enhancement" if it is kept though.

**Comment by BaptisteFrancois** [comment]
Looks like I have not the right to label an issue.

**Comment by jetuk** [comment]
Labelled, thanks.
